### PR TITLE
Nested folder picker: Move options into overlay

### PR DESCRIPTION
**Pull request by ashharrison90**
**What is this feature?**

- test by enabling the `nestedFolderPicker` toggle
- moves the folder picker options into an overlay
  - initial button style is very WIP, just wanted something that worked 😅 
  - restricts the number of rows to either the number of items or 6.5, whichever is smaller. this prevents ugly empty space when only a couple of results.
  - adjusts some styling to match how our other selects render an overlay
  - nested folder picker now also needs to be passed the folder title so it can display it correctly without making an api call
  
https://github.com/grafana/grafana/assets/20999846/19f2bc43-2c15-4a61-b046-236ab57955bb

**Why do we need this feature?**

- better UX

**Who is this feature for?**

- everyone using nested folders!

**Which issue(s) does this PR fix?**:

For https://github.com/grafana/grafana/issues/65604

**Special notes for your reviewer:**

- this **does not** handle keyboard support. there will be a follow up PR for that.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by amy-super** [comment]
I think this looks amazing! It works much better and is a cleaner interaction than the other designs. 

One quick bit of feedback, and then the todos. 

Feedback: 
I think we could use a bit more padding between the chevron and the text, if only to create a cleaner separation between the click targets. We achieve that in browse view by putting the folder/dashboard icon between the chevron and the name. Since everything here is a folder, I think a bit more space (4 px?) would help. 
Screenshot 2023-07-05 at 1 46 59 PM

Todo:
- Figure out how to add a new folder in this design
- Choose a "starting" element that isn't a button. Or maybe it is!
- Keyboard stuff as mentioned in our discussion

**Comment by tskarhed** [comment]
Should empty folders have the chevron next to them?

File: public/app/features/browse-dashboards/components/BrowseActions/MoveModal.tsx [hunk 1]
```diff
@@ -21,19 +21,26 @@ export interface Props {
...
+  const [moveTarget, setMoveTarget] = useState<{
+    uid?: string;
+    title?: string;
+  }>({});
   const [isMoving, setIsMoving] = useState(false);
   const selectedFolders = Object.keys(selectedItems.folder).filter((uid) => selectedItems.folder[uid]);
 
   const handleFolderChange = (newFolder: FolderChange) => {
-    setMoveTarget(newFolder.uid === ROOT_FOLDER ? '' : newFolder.uid);
+    // setMoveTarget(newFolder.uid === ROOT_FOLDER ? '' : newFolder.uid);
```
Comment (tskarhed) [inline public/app/features/browse-dashboards/components/BrowseActions/MoveModal.tsx:32]:
```suggestion
```

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.test.tsx
Comment (tskarhed) [inline public/app/core/components/NestedFolderPicker/NestedFolderPicker.test.tsx:1]: Why do your tests always feel like a cup of hot chocolate and a blanket? :1st_place_medal:

**Comment by ashharrison90** [comment]
> Should empty folders have the chevron next to them?

the problem is until we click the chevron, we have no way of knowing whether a folder is empty or not. this behaviour is similar to mac's finder, see https://raintank-corp.slack.com/archives/C025WTVRBSN/p1686815545012439?thread_ts=1686755827.550399&cid=C025WTVRBSN for a video

File: public/app/core/components/NestedFolderPicker/NestedFolderList.tsx [hunk 2]
```diff
@@ -172,7 +175,9 @@ const getStyles = (theme: GrafanaTheme2) => {
       display: 'flex',
       position: 'relative',
       alignItems: 'center',
-      borderTop: `solid 1px ${theme.components.input.borderColor}`,
+      [':not(:first-child)']: {
+        borderTop: `solid 1px ${theme.components.input.borderColor}`,
```
Comment (tskarhed) [inline public/app/core/components/NestedFolderPicker/NestedFolderList.tsx:179]: I feel like the contrast for this color is a bit high. It probably works better on a greyer background.

**Review by eledobleefe — APPROVED** [review]
I checked this locally and LGTM 👌

**Review by JoaoSilvaGrafana — COMMENTED** [review]
This looks cool! 👏 

I was testing it out and found a couple of things that we could change:

1. We don't have a way to truncate long names yet, either in the button or the folder list:
Screenshot 2023-07-11 at 13 15 42
Screenshot 2023-07-11 at 13 15 55

2. When you click "clear" after searching for something, it just closes the picker. I'm assuming because it recognizes it as a click outside the folder list?
3. I don't know if this is the right interaction or not, but when you have a folder selected and then you open the list and select the same folder again, nothing happens. I wonder if it should just close the list then as having something selected?

**Review by joshhunt — COMMENTED** [review]
UX wise, I think this is in a good place to iterate from.

Just a few small minor nits, but I would really like it if we kept `value` to just the UID if possible...

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx [hunk 3]
```diff
@@ -122,66 +126,118 @@ export function NestedFolderPicker({ value, onChange }: NestedFolderPickerProps)
...
+    offset: [0, 0],
+    trigger: 'click',
+    onVisibleChange: (value: boolean) => {
+      setOverlayOpen(value);
+    },
+  });
+
+  useEffect(() => {
+    if (!overlayOpen) {
+      setSearch('');
```
[the remark this answers is not in the document]
Comment (ashharrison90) [reply public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:147]: oh yeah, idk how i missed that, that's much cleaner 😅

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx [hunk 4]
```diff
@@ -13,23 +13,27 @@ import { queryResultToViewItem } from 'app/features/search/service/utils';
...
 interface NestedFolderPickerProps {
-  value?: FolderUID | undefined;
+  value: {
+    title?: string;
+    uid?: FolderUID;
+  };
   // TODO: think properly (and pragmatically) about how to communicate moving to general folder,
   // vs removing selection (if possible?)
-  onChange?: (folderUID: FolderChange) => void;
+  onChange?: (folder: FolderChange) => void;
```
[the remark this answers is not in the document]
Comment (ashharrison90) [reply public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:28]: currently the `onChange` in the consuming components is shared between this and the old folder picker, so it needs to stay like this for now until we remove the old folder picker. we _could_ write separate `onChange` functions in each of the consuming components... wdyt? 🤔

**Comment by ashharrison90** [comment]
> This looks cool! 👏
> 
> I was testing it out and found a couple of things that we could change:
> 
> 1. We don't have a way to truncate long names yet, either in the button or the folder list:
> 
> Screenshot 2023-07-11 at 13 15 42 Screenshot 2023-07-11 at 13 15 55
> 2. When you click "clear" after searching for something, it just closes the picker. I'm assuming because it recognizes it as a click outside the folder list?
> 3. I don't know if this is the right interaction or not, but when you have a folder selected and then you open the list and select the same folder again, nothing happens. I wonder if it should just close the list then as having something selected?

@JoaoSilvaGrafana agree with all your points. 👍  will tackle them in follow up PRs (if that's ok 🙏) as i'm pretty sure they're all gonna be affected by the changes we need to make for keyboard handling e.g. i doubt we're gonna keep the fact that it's currently a radio button list, which is responsible for the interaction in point 3. this will also probably change the dom structure so i don't wanna look at truncation until that's settled a bit, etc.

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:28]: oh. you didn't actually change this. never mind.

**Review by joshhunt — APPROVED** [review]

**Review by JoaoSilvaGrafana — APPROVED** [review]

**Pull request merged**